website/www/site/content/en/blog/unit-testing-blog.md

```diff
@@ -0,0 +1,149 @@
+
+
+Here are the corresponding tests for both pipelines:
+
+    # The following packages are imported for unit testing.
+    import unittest
+    import apache_beam as beam
+
+
+    @unittest.skipIf(HttpError is None, 'GCP dependencies are not installed')
+    class TestBeam(unittest.TestCase):
+
+    # This test corresponds to pipeline p1, and is written to confirm the compute_square function works as intended.
+        def test_compute_square(self):
+            numbers=[1,2,3]
+
+
+        with TestPipeline() as p:
+            output = p | beam.Create([1,2,3])
+                       | beam.Map(compute_square)
+        assert_that(output, equal_to([1,4,9]))
+
+
+
+    # This test corresponds to pipeline p2, and is written to confirm that our map operation works as intended.
+ 	def test_strip_map(self):
+		strings= [' Strawberry   \n','   Carrot   \n','   Eggplant   \n']
+		with TestPipeline() as p:
+			output = p | beam.Create(strings)
+                               | beam.Map(str.strip)
+        assert_that(output,['Strawberry','Carrot','Eggplant'])
+
+
+
+The following cover other testing best practices:
+
+1. Test all error messages you raise.
+2. Cover any edge cases that might be present in your data.
+3. Notice that in pipeline 1, we could have written the `beam.Map` step as  `beam.Map(lambda x: x**2)`, instead of `beam.Map(compute_square)`. The latter (separating the lambda into a helper function) is the recommended approach for more testable code, as changes to the function would be modularized.
+4. Use the `assert_that` statement to ensure that PCollection values match up correctly, such as the following example:
+
+
+      class TestBeam(unittest.TestCase):
+          def test_custom_function(self):
+              with TestPipeline() as p:
+                input = p | beam.ParDo(custom_function(("1","2","3"))
+              assert_that(input, equal_to(["1","2","3"]))
+
+
+5. If needed, use mocking to mock any API calls that might be present in your ParDo function. The purpose of mocking is to test your functionality extensively, even if this testing requires a specific response from an API call.
+
+The following snippet is based off of example pipeline 1, from the top of this blog post.
+
+      # We import the mock package for mocking functionality
+      import mock
+
+
+      @mock.patch.object(CustomFunction, 'get_record')
+      def test_error_message_wrong_length(self, get_record):
+        record = ["field1","field2",...]
+        get_record.return_value = record
+        with self.assertRaisesRegex(ValueError,
+                                    "Length of record does not match expected length'"):
+            p = apache_beam.Pipeline()
+            result = p | beam.ParDo(custom_function(x))
+        result
+
+
+
+
+
+
+For more pointed guidance on testing on Beam/Dataflow, see the [Google Cloud documentation](https://cloud.google.com/dataflow/docs/guides/develop-and-test-pipelines).
```

Comment:
Could we also point to some example tests here as well? The beam repo is probably the best place for example PTransform tests. For example, we have quite extensive RunInference tests - https://github.com/apache/beam/blob/736cf50430b375d32093e793e1556567557614e9/sdks/python/apache_beam/ml/inference/base_test.py#L262 - and we could add similarly well tested transforms in Java and Go